source/common/network/cidr_range.cc

```diff
@@ -0,0 +1,177 @@
+        if ((ptr->ip()->version() == IpVersion::v6 && length64 <= 128) ||
+            (ptr->ip()->version() == IpVersion::v4 && length64 <= 32)) {
+          return construct(ptr, static_cast<uint32_t>(length64));
+        }
+      }
+    }
+  }
+  return CidrRange(nullptr, -1);
+}
+
+// static
+InstancePtr CidrRange::truncateIpAddressAndLength(const InstancePtr& address, int* length_io) {
+  int length = *length_io;
+  if (address == nullptr || length < 0 || address->type() != Type::Ip) {
+    *length_io = -1;
+    return nullptr;
+  }
+  switch (address->ip()->version()) {
+  case IpVersion::v4: {
+    if (length >= 32) {
+      // We're using all of the bits, so don't need to create a new address instance.
+      *length_io = 32;
+      return address;
+    } else if (length == 0) {
+      // Create an Ipv4Instance with only a port, which will thus have the any address.
+      // TODO( Consider adding a Utility::AnyIpv4Address() static method to return such an instance.
+      return InstancePtr(new Ipv4Instance(uint32_t(0)));
+    }
+    // Need to mask out the unused bits, and create an Ipv4Instance with this address.
+    uint32_t ip4 = ntohl(address->ip()->ipv4()->address());
+    ip4 &= ~0U << (32 - length);
+    sockaddr_in sa4;
+    sa4.sin_family = AF_INET;
+    sa4.sin_port = htons(0);
+    sa4.sin_addr.s_addr = htonl(ip4);
+    return InstancePtr(new Ipv4Instance(&sa4));
+  }
+
+  case IpVersion::v6: {
+    if (length >= 128) {
+      // We're using all of the bits, so don't need to create a new address instance.
+      *length_io = 128;
+      return address;
+    } else if (length == 0) {
+      // Create an Ipv6Instance with only a port, which will thus have the any address.
+      // TODO Consider adding an Any6() static method to return such an instance.
+      return InstancePtr(new Ipv6Instance(uint32_t(0)));
+    }
+    // We need to mask out the unused bits, but we don't have a uint128_t available.
+    // However, we know that the array returned by Ipv6::address() represents the address
+    // in network order (bit 7 of byte 0 is the high-order bit of the address), so we
+    // simply need to keep the leading length bits of the array, and get rid of the rest.
+    sockaddr_in6 sa6;
+    sa6.sin6_family = AF_INET6;
+    sa6.sin6_port = htons(0);
+    std::array<uint8_t, 16> ip6 = address->ip()->ipv6()->address();
+    for (int i = 0; i < 16; ++i) {
+      if (length == 0) {
+        // We've retained all the bits we need, the remaining are all zero.
+        sa6.sin6_addr.s6_addr[i] = 0;
+      } else if (length < 8) {
+        // We're retaining only the high-order length bits of this byte.
+        sa6.sin6_addr.s6_addr[i] = ip6[i] & (0xff << (8 - length));
+        length = 0;
+      } else {
+        // We're keeping all of these bits.
+        sa6.sin6_addr.s6_addr[i] = ip6[i];
+        length -= 8;
+      }
+    }
+    return InstancePtr(new Ipv6Instance(sa6));
+  }
+  }
+  return nullptr; // UNREACHED
```

Comment:
Sure, that's fine too. Feel to add a NOT_REACHED macro near NOT_IMPLEMENTED that also crashes and does other stuff.